### PR TITLE
[WIP] refactor/plugin system

### PR DESCRIPTION
**Pull request by efegurkan**
Rewrite and rearrange utilities into plugins.

-Create sqip.js to run plugins over the input file
-Turn SVG, SVGO, primitive and base64 as plugins
-Move loadSVG into utils/helpers

### CODE REVIEW

**Comment by efegurkan** [comment]
I started experimenting with a possible implementation of the plugin system for a while ago. I will reintegrate after #42 has been merged.

@axe312ger Feedback would be awesome since you have better experience with node compared to me :) Also I am pretty sure you have something in your mind as well.

File: src/plugins/base64.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+const encodeBase64 = rawSVG => Buffer.from(rawSVG).toString('base64')
```
Comment (axe312ger) [inline src/plugins/base64.js]: I wonder if we should remove this completely. Base64 is just bad for compression. See data here:

https://github.com/gatsbyjs/gatsby/pull/4205#issuecomment-368348758

Solution: https://www.npmjs.com/package/mini-svg-data-uri

File: src/plugins/svgo.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+// SVGO with settings for maximum compression to optimize the Primitive-generated SVG
+class SVGOPlugin {
+  constructor(options) {
+    this.options = options || {}
+  }
+  apply(svg) {
+    return this.optimize(svg)
+  }
+  optimize(svg) {
+    const { multipass = true, floatPrecision = 1 } = this.options
```
Comment (axe312ger) [inline src/plugins/svgo.js]: These default options should be in the constructor IMHO

File: src/sqip.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+const fs = require('fs')
+const path = require('path')
+
+const SVGPlugin = require('./plugins/svg')
+const SVGOPlugin = require('./plugins/svgo')
+const PrimitivePlugin = require('./plugins/primitive')
+const Base64EncodePlugin = require('./plugins/base64')
```
Comment (axe312ger) [inline src/sqip.js]: i wonder if we should at least export them as separate packages with something like https://github.com/lerna/lerna

File: src/plugins/base64.js (hunk 1 above)
Comment (efegurkan) [reply src/plugins/base64.js]: @technopagan had something about that I can agree on that <3 also the wrapping we do, not sure if it is useful at all.

File: src/sqip.js (hunk 3 above)
Comment (efegurkan) [reply src/sqip.js]: Good point, we are actually using it for `melody` at trivago.

File: src/plugins/base64.js (hunk 1 above)
Comment (axe312ger) [reply src/plugins/base64.js]: Actually I expected the CLI to spit out a SVG file, not an Image tag 🙈 

We could keep base64 as plugin and use the mini data uri's as default for encoding.

**Comment by efegurkan** [comment]
@axe312ger @technopagan Hey I am sorry I cannot find any more time to continue on it currently. A lot has been happening since last 6 months (Personal stuff, work change etc.). I feel like I am blocking everything. Please feel free to take over this, or build a new one. I will still try to do contribute later, but cannot find free time anymore.

Sorry :man_shrugging:

**Comment by axe312ger** [comment]
@hafizalfaza this is something you could have a look at. Rebase, fix conflicts and then check how much of @efegurkan code we can use. I guess it is in a pretty good state.

**Comment by axe312ger** [comment]
code went into core via #55

**Pull request closed without merging**